### ISSUE TITLE
Test Executor sporadically fails with exit code 1 after tests have been executed

### ISSUE THREAD

**Issue by marcphilipp** [gradle/gradle#11195]
The tesk task sporadically fails after tests have been executed with an error message like the following:

```
Process 'Gradle Test Executor 1' finished with non-zero exit value 1
This problem might be caused by incorrect test process configuration.
Please refer to the test execution section in the User Manual at https://docs.gradle.org/5.6.2/userguide/java_testing.html#sec:test_execution
```

There were no logs in `~/.gradle/workers`.

### Expected Behavior

The `test` task finishes successfully.

### Current Behavior

It fails.

### Context

This is hard to reproduce since it only happens sometimes. After some initial discussion with @big-guy and @ghale, we suspect that something is failing during shutdown of the worker. However, we don't see any logs because the connection has already been stopped so `unrecoverableErrorHandler` is not used.

### Steps to Reproduce 

Run `:spring-kafka:test` and repeat until it fails. 😉 

### Your Environment

Build scan URL: https://ge.spring.io/s/uie55fwwto53i

**Comment by prachisr** [comment]
We are facing the same issue. We are trying to run integrationTest task that runs test that uses EmbeddedKafka.
Even when the test are passing we are getting the task failure.
It gives me a similar error:
 
What went wrong:
Execution failed for task ':integrationTest'.
> Process 'Gradle Test Executor 19' finished with non-zero exit value 1
  This problem might be caused by incorrect test process configuration.
  Please refer to the test execution section in the User Manual at https://docs.gradle.org/6.0.1/userguide/java_testing.html#sec:test_execution

Any help would be much appreciated.

**Comment by pwamej** [comment]
+1 for this, I'm experiencing the same problem in tests that set up Spring and Kafka.

**Comment by bmuskalla** [comment]
My guess would be that this is related to `embeddedkafka` in that it calls `System.exit` at some point if it is not shutdown properly during a test (would actually be great to intercept this in gradle and properly warn the user). There is a related issue on the `embeddedkafka` side: https://github.com/spring-projects/spring-kafka/issues/345

**Comment by IndeedSi** [comment]
Same issue here. We are using Spring-boot and Spring-kafka libraries. 
In our repo, it happens about once in five runs.

**Comment by pwamej** [comment]
We've managed to get rid of the problem by setting halt procedure for `org.apache.kafka.common.utils.Exit`:
```java
import org.apache.kafka.common.utils.Exit

Exit.setHaltProcedure(new Exit.Procedure() {
    @Override
    void execute(int statusCode, String message) {
        throw new RuntimeException(
                String.format("halting requested statusCode %d, message %s", statusCode, message))
    }
})
```
throwing Exception instead of exitting the program allows Gradle to handle it. It's a dirty workaround, but maybe it will lead someone to correct solution.

**Comment by garyrussell** [comment]
Thanks @pwamej - someone reported a similar work-around [here](https://github.com/spring-projects/spring-kafka/issues/194#issuecomment-612875646).

But I didn't join the dots to figure out this was causing the sporadic Gradle failure on the Spring Kafka build.

@marcphilipp I believe you can close this.

**Comment by marcphilipp** [comment]
Just to get this straight: the reason was that Kafka called `System.exit(1)`?

**Comment by garyrussell** [comment]
@marcphilipp That's my understanding, yes (but I need to do a bit more investigation).

spring-kafka-test provides support for an embedded Kafka server (manages some scala classes provided by Kafka) and it appears there is a race condition during shutdown. The issue I referenced was about missing files but it appears that if the shutdown occurs too quickly, the JVM is exited.

I wish I had connected the dots between the missing files issue and this issue earlier, for which I apologize.

Bottom line is spring-kafka needs to disable the `System.exit()` and it looks like there are hooks for us to do that.

Perhaps to help in future situations, Gradle could emit

`Test Executor sporadically fails with exit code 1 after tests have been executed - perhaps something called System.exit()`

??

(I won't need this, but it might help others 😄 ).

**Comment by marcphilipp** [comment]
@big-guy Has this ever come up before? Could we install a default security manager to prevent code from calling `System.exit()`?

**Comment by stmlange** [comment]
Any updates on this? I think it would be really useful if gradle would provide an option/configuration to install a default security manager to prevent code from calling `System.exit()`. Especially after reading the note about the security manager in https://docs.gradle.org/current/userguide/java_testing.html#sec:test_execution makes me wish that gradle could take of that.

(As a side note also discovered a test that was using `System.exit` by accident, took quite some time to find the root cause of this.)

**Comment by shu-bham** [comment]
I'm getting this issue on my Spring boot project while running `MockWebServer` tests

**Comment by marcphilipp** [comment]
@napstr Currently nothing prevents you from calling `System.exit` from test code.
